[SIMD] Intel support for saturating integer arithmetic operations

hyjorc1 opened the pull request:
#### 5a0f6d496df12411613df873460a8b0717a65459
<pre>
[SIMD] Intel support for saturating integer arithmetic operations
<a href="https://bugs.webkit.org/show_bug.cgi?id=248862">https://bugs.webkit.org/show_bug.cgi?id=248862</a>
rdar://103061350

Reviewed by Yusuke Suzuki.

Add WASM SIMD saturating integer arithmetic operations.
<a href="https://github.com/WebAssembly/simd/blob/main/proposals/simd/SIMD.md#saturating-integer-arithmetic">https://github.com/WebAssembly/simd/blob/main/proposals/simd/SIMD.md#saturating-integer-arithmetic</a>

* Source/JavaScriptCore/assembler/MacroAssemblerX86_64.h:
(JSC::MacroAssemblerX86_64::vectorMax):
(JSC::MacroAssemblerX86_64::vectorMin):
(JSC::MacroAssemblerX86_64::vectorAbs):
(JSC::MacroAssemblerX86_64::vectorAddSat):
(JSC::MacroAssemblerX86_64::vectorSubSat):
(JSC::MacroAssemblerX86_64::vectorAvgRound):
(JSC::MacroAssemblerX86_64::vectorMulSat):
* Source/JavaScriptCore/assembler/X86Assembler.h:
(JSC::X86Assembler::paddsb_rr):
(JSC::X86Assembler::vpaddsb_rr):
(JSC::X86Assembler::paddusb_rr):
(JSC::X86Assembler::vpaddusb_rr):
(JSC::X86Assembler::paddsw_rr):
(JSC::X86Assembler::vpaddsw_rr):
(JSC::X86Assembler::paddusw_rr):
(JSC::X86Assembler::vpaddusw_rr):
(JSC::X86Assembler::psubsb_rr):
(JSC::X86Assembler::vpsubsb_rr):
(JSC::X86Assembler::psubusb_rr):
(JSC::X86Assembler::vpsubusb_rr):
(JSC::X86Assembler::psubsw_rr):
(JSC::X86Assembler::vpsubsw_rr):
(JSC::X86Assembler::psubusw_rr):
(JSC::X86Assembler::vpsubusw_rr):
(JSC::X86Assembler::pmaxsb_rr):
(JSC::X86Assembler::vpmaxsb_rr):
(JSC::X86Assembler::pmaxsw_rr):
(JSC::X86Assembler::vpmaxsw_rr):
(JSC::X86Assembler::pmaxsd_rr):
(JSC::X86Assembler::vpmaxsd_rr):
(JSC::X86Assembler::pmaxub_rr):
(JSC::X86Assembler::vpmaxub_rr):
(JSC::X86Assembler::pmaxuw_rr):
(JSC::X86Assembler::vpmaxuw_rr):
(JSC::X86Assembler::pmaxud_rr):
(JSC::X86Assembler::vpmaxud_rr):
(JSC::X86Assembler::pminsb_rr):
(JSC::X86Assembler::vpminsb_rr):
(JSC::X86Assembler::pminsw_rr):
(JSC::X86Assembler::vpminsw_rr):
(JSC::X86Assembler::pminsd_rr):
(JSC::X86Assembler::vpminsd_rr):
(JSC::X86Assembler::pminub_rr):
(JSC::X86Assembler::vpminub_rr):
(JSC::X86Assembler::pminuw_rr):
(JSC::X86Assembler::vpminuw_rr):
(JSC::X86Assembler::pminud_rr):
(JSC::X86Assembler::vpminud_rr):
(JSC::X86Assembler::pavgb_rr):
(JSC::X86Assembler::vpavgb_rr):
(JSC::X86Assembler::pavgw_rr):
(JSC::X86Assembler::vpavgw_rr):
(JSC::X86Assembler::pabsb_rr):
(JSC::X86Assembler::vpabsb_rr):
(JSC::X86Assembler::pabsw_rr):
(JSC::X86Assembler::vpabsw_rr):
(JSC::X86Assembler::pabsd_rr):
(JSC::X86Assembler::vpabsd_rr):
(JSC::X86Assembler::vpxor_rr):
(JSC::X86Assembler::vpsubq_rr):
(JSC::X86Assembler::vblendvpd_rr):
(JSC::X86Assembler::vpmulhrsw_rr):
(JSC::X86Assembler::vpcmpeqw_rr):
(JSC::X86Assembler::X86InstructionFormatter::SingleInstructionBufferWriter::memoryModRM):
* Source/JavaScriptCore/b3/air/AirOpcode.opcodes:
* Source/JavaScriptCore/wasm/WasmAirIRGenerator.cpp:
(JSC::Wasm::AirIRGenerator::addSIMDSwizzle):
(JSC::Wasm::AirIRGenerator::addSIMDV_VV):

Canonical link: <a href="https://commits.webkit.org/257468@main">https://commits.webkit.org/257468@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/77a1d613064da633bbb56d3b4412b676d881de7f

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [❌ 🧪 style](https://ews-build.webkit.org/#/builders/6/builds/99121 "check-webkit-style running") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/77/builds/8326 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/43/builds/32259 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/8/builds/108515 "Built successfully") | [✅ 🛠 🧪 win](https://ews-build.webkit.org/#/builders/10/builds/168757 "Built successfully and passed tests") 
| | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/76/builds/8880 "Built successfully") | [  ~~🛠 mac-AS-debug~~](https://ews-build.webkit.org/#/builders/61/builds/85671 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/36/builds/91628 "Built successfully") | [  ~~🛠 wincairo~~](https://ews-build.webkit.org/#/builders/12/builds/106452 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/19/builds/104858 "Passed tests") | [  ~~🧪 ios-wk2~~](https://ews-build.webkit.org/#/builders/78/builds/6719 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 api-mac~~](https://ews-build.webkit.org/#/builders/3/builds/90289 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/35/builds/33727 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [  ~~🧪 api-ios~~](https://ews-build.webkit.org/#/builders/9/builds/88536 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 mac-wk1~~](https://ews-build.webkit.org/#/builders/73/builds/21634 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 api-gtk~~](https://ews-build.webkit.org/#/builders/34/builds/76587 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| [✅ 🛠 🧪 jsc](https://ews-build.webkit.org/#/builders/1/builds/89832 "Built successfully and passed tests") | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/81/builds/2217 "Built successfully") | [  ~~🧪 mac-wk2~~](https://ews-build.webkit.org/#/builders/70/builds/23150 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 jsc-armv7](https://ews-build.webkit.org/#/builders/38/builds/85661 "Built successfully") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/82/builds/2109 "Built successfully") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/61/builds/85671 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [✅ 🧪 jsc-armv7-tests](https://ews-build.webkit.org/#/builders/46/builds/28851 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/74/builds/5144 "Built successfully and passed tests") | [  ~~🛠 watch~~](https://ews-build.webkit.org/#/builders/80/builds/7071 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 mac-wk2-stress~~](https://ews-build.webkit.org/#/builders/62/builds/42628 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 jsc-mips](https://ews-build.webkit.org/#/builders/37/builds/88523 "Built successfully") | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/79/builds/3526 "Built successfully") | | [✅ 🧪 jsc-mips-tests](https://ews-build.webkit.org/#/builders/45/builds/19815 "Passed tests") | 
<!--EWS-Status-Bubble-End-->

Constellation commented on Source/JavaScriptCore/assembler/MacroAssemblerX86_64.h:
```diff
@@ -2353,29 +2353,131 @@ class MacroAssemblerX86_64 : public MacroAssemblerX86Common {
+                if (simdInfo.signMode == SIMDSignMode::Signed)
+                    m_assembler.pmaxsw_rr(right, dest);
+                else {
+                    if (supportsFloatingPointRounding())
```
Add `supportsSSE4_1()` and use it instead. `supportsFloatingPointRounding` should be used only for rouding.

Constellation commented on Source/JavaScriptCore/assembler/MacroAssemblerX86_64.h:
```diff
@@ -2353,29 +2353,131 @@ class MacroAssemblerX86_64 : public MacroAssemblerX86Common {
+                if (left != dest)
+                    m_assembler.movapd_rr(left, dest);
+                if (simdInfo.signMode == SIMDSignMode::Signed) {
+                    if (supportsFloatingPointRounding())
```
Ditto.

Constellation commented on Source/JavaScriptCore/assembler/MacroAssemblerX86_64.h:
```diff
@@ -2353,29 +2353,131 @@ class MacroAssemblerX86_64 : public MacroAssemblerX86Common {
+                if (left != dest)
+                    m_assembler.movapd_rr(left, dest);
+                if (simdInfo.signMode == SIMDSignMode::Signed) {
+                    if (supportsFloatingPointRounding())
```
Ditto.

Constellation commented on Source/JavaScriptCore/assembler/MacroAssemblerX86_64.h:
```diff
@@ -2353,29 +2353,131 @@ class MacroAssemblerX86_64 : public MacroAssemblerX86Common {
+                if (simdInfo.signMode == SIMDSignMode::Signed)
+                    m_assembler.pminsw_rr(right, dest);
+                else {
+                    if (supportsFloatingPointRounding())
```
Ditto.

Constellation commented on Source/JavaScriptCore/assembler/MacroAssemblerX86_64.h:
```diff
@@ -2353,29 +2353,131 @@ class MacroAssemblerX86_64 : public MacroAssemblerX86Common {
+                if (left != dest)
+                    m_assembler.movapd_rr(left, dest);
+                if (simdInfo.signMode == SIMDSignMode::Signed) {
+                    if (supportsFloatingPointRounding())
```
Ditto.

Constellation commented on Source/JavaScriptCore/assembler/MacroAssemblerX86_64.h:
```diff
@@ -2430,11 +2532,48 @@ class MacroAssemblerX86_64 : public MacroAssemblerX86Common {
+                m_assembler.vpxor_rr(dest, dest, dest);
+                m_assembler.vpsubq_rr(input, dest, dest);
+                m_assembler.vblendvpd_rr(input, dest, input, dest);
+            } else if (supportsFloatingPointRounding()) {
```
Ditto.

Constellation commented on Source/JavaScriptCore/assembler/MacroAssemblerX86_64.h:
```diff
@@ -2353,29 +2353,131 @@ class MacroAssemblerX86_64 : public MacroAssemblerX86Common {
+                if (left != dest)
+                    m_assembler.movapd_rr(left, dest);
+                if (simdInfo.signMode == SIMDSignMode::Signed) {
+                    if (supportsFloatingPointRounding())
```
Ditto.

Constellation reviewed (approved):
r=me with comments.

hyjorc1 committed:
[SIMD] Intel support for saturating integer arithmetic operations
https://bugs.webkit.org/show_bug.cgi?id=248862
rdar://103061350

Reviewed by Yusuke Suzuki.

Add WASM SIMD saturating integer arithmetic operations.
https://github.com/WebAssembly/simd/blob/main/proposals/simd/SIMD.md#saturating-integer-arithmetic

* Source/JavaScriptCore/assembler/MacroAssemblerX86_64.h:
(JSC::MacroAssemblerX86_64::vectorMax):
(JSC::MacroAssemblerX86_64::vectorMin):
(JSC::MacroAssemblerX86_64::vectorAbs):
(JSC::MacroAssemblerX86_64::vectorAddSat):
(JSC::MacroAssemblerX86_64::vectorSubSat):
(JSC::MacroAssemblerX86_64::vectorAvgRound):
(JSC::MacroAssemblerX86_64::vectorMulSat):
* Source/JavaScriptCore/assembler/X86Assembler.h:
(JSC::X86Assembler::paddsb_rr):
(JSC::X86Assembler::vpaddsb_rr):
(JSC::X86Assembler::paddusb_rr):
(JSC::X86Assembler::vpaddusb_rr):
(JSC::X86Assembler::paddsw_rr):
(JSC::X86Assembler::vpaddsw_rr):
(JSC::X86Assembler::paddusw_rr):
(JSC::X86Assembler::vpaddusw_rr):
(JSC::X86Assembler::psubsb_rr):
(JSC::X86Assembler::vpsubsb_rr):
(JSC::X86Assembler::psubusb_rr):
(JSC::X86Assembler::vpsubusb_rr):
(JSC::X86Assembler::psubsw_rr):
(JSC::X86Assembler::vpsubsw_rr):
(JSC::X86Assembler::psubusw_rr):
(JSC::X86Assembler::vpsubusw_rr):
(JSC::X86Assembler::pmaxsb_rr):
(JSC::X86Assembler::vpmaxsb_rr):
(JSC::X86Assembler::pmaxsw_rr):
(JSC::X86Assembler::vpmaxsw_rr):
(JSC::X86Assembler::pmaxsd_rr):
(JSC::X86Assembler::vpmaxsd_rr):
(JSC::X86Assembler::pmaxub_rr):
(JSC::X86Assembler::vpmaxub_rr):
(JSC::X86Assembler::pmaxuw_rr):
(JSC::X86Assembler::vpmaxuw_rr):
(JSC::X86Assembler::pmaxud_rr):
(JSC::X86Assembler::vpmaxud_rr):
(JSC::X86Assembler::pminsb_rr):
(JSC::X86Assembler::vpminsb_rr):
(JSC::X86Assembler::pminsw_rr):
(JSC::X86Assembler::vpminsw_rr):
(JSC::X86Assembler::pminsd_rr):
(JSC::X86Assembler::vpminsd_rr):
(JSC::X86Assembler::pminub_rr):
(JSC::X86Assembler::vpminub_rr):
(JSC::X86Assembler::pminuw_rr):
(JSC::X86Assembler::vpminuw_rr):
(JSC::X86Assembler::pminud_rr):
(JSC::X86Assembler::vpminud_rr):
(JSC::X86Assembler::pavgb_rr):
(JSC::X86Assembler::vpavgb_rr):
(JSC::X86Assembler::pavgw_rr):
(JSC::X86Assembler::vpavgw_rr):
(JSC::X86Assembler::pabsb_rr):
(JSC::X86Assembler::vpabsb_rr):
(JSC::X86Assembler::pabsw_rr):
(JSC::X86Assembler::vpabsw_rr):
(JSC::X86Assembler::pabsd_rr):
(JSC::X86Assembler::vpabsd_rr):
(JSC::X86Assembler::vpxor_rr):
(JSC::X86Assembler::vpsubq_rr):
(JSC::X86Assembler::vblendvpd_rr):
(JSC::X86Assembler::vpmulhrsw_rr):
(JSC::X86Assembler::vpcmpeqw_rr):
(JSC::X86Assembler::X86InstructionFormatter::SingleInstructionBufferWriter::memoryModRM):
* Source/JavaScriptCore/b3/air/AirOpcode.opcodes:
* Source/JavaScriptCore/wasm/WasmAirIRGenerator.cpp:
(JSC::Wasm::AirIRGenerator::addSIMDSwizzle):
(JSC::Wasm::AirIRGenerator::addSIMDV_VV):

Canonical link: https://commits.webkit.org/257468@main
